providers/base/bin/check_audio_deamon.sh

```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+if [ -z "$(command -v pactl)" ]; then
+    if [ -n "$(command -v pw-cli)" ] ;then
+        echo "Daemon is pipewire, use pipewire function"
+        exit 0
+    else
+        # shouldn't be here,
+        # if requires: package.name in ['pulseaudio-utils', 'pipewire']
+        echo "No pipewire or pulseaudio installed!! Stop test"
+        exit 2
+    fi
+elif pactl info | grep -iq pipewire ; then
+    echo "Daemon is pipewire, use pipeiwre function"
+    exit 0
+else
+    echo "Daemon is pulseaudio, use pulseaudio function"
+    exit 1
+fi
```

Comment:
The logic here confuse me for 3 reasons:
1. The number of exit code has [meaning](https://pressbooks.senecacollege.ca/uli101/chapter/exit-status-in-shell-scripting/), `1`: general error, `2`: missing attribute. One of your case is calling pulseaudio and exit with `1`, this is not the common way to use it, would leds to misunderstanding.
2. In the upper layer: audio/job.pxu, you wrote as below, which is not right, because if `check_audio_deamon.sh` find no cmd and exit with code: 2, it will return to the `else` section in job.pxu and still trying to call audio_settings.py. This is why we shouldn't write complex sh in job.pxu. Unit test won't be able to catch errors if any.
3. Why need `EXIT_CODE=$?` here? If I understand correct, if any of the script here failed at something will cause execution stop and exit with `1`, and if not, it would be `0` defaultly. Is there a reason to point that out in code?
4. I'm not sure the echo msg is necessary, since the lower layer of script should have enough msg to help user know what is exactly happening(for both pass and error). I might be wrong about this, but current workflow is Jenkins -> TF -> Checkbox -> testplan -> jobs, each stage should only provide the msg relates to itself, otherwise, the log will be tremendous, and not helping us to understand what's happening.
```
  if check_audio_deamon.sh ; then
    show_default_audio_device.sh
    pipewire_utils.py gst -t 2 'audiotestsrc wave=sine freq=512 ! audioconvert ! audioresample ! autoaudiosink'
    EXIT_CODE=$?
  else
    audio_settings.py store --file="$PLAINBOX_SESSION_SHARE"/pulseaudio_settings
    audio_settings.py set --device=pci --volume=50
    gst_pipeline_test.py -t 2 'audiotestsrc wave=sine freq=512 ! audioconvert ! audioresample ! autoaudiosink'
    EXIT_CODE=$?
    audio_settings.py restore --file="$PLAINBOX_SESSION_SHARE"/pulseaudio_settings
  fi
```

Could you please let me know why it can not be as simple as below? Am I missing some thing here?
```
if command -v pw-cli; then
    show_default_audio_device.sh
    pipewire_utils.py gst -t 2 'audiotestsrc wave=sine freq=512 ! audioconvert ! audioresample ! autoaudiosink'

elif command -v pactl; then
    audio_settings.py store --file="$PLAINBOX_SESSION_SHARE"/pulseaudio_settings
    audio_settings.py set --device=pci --volume=50
    gst_pipeline_test.py -t 2 'audiotestsrc wave=sine freq=512 ! audioconvert ! audioresample ! autoaudiosink'
    audio_settings.py restore --file="$PLAINBOX_SESSION_SHARE"/pulseaudio_settings

else
    echo "Test failed: cannot found pipewire or pulseaudio"
    exit 1
```

Reply:
For 2. the job has defined `package.name in ['pulseaudio-utils', 'pipewire']`, therefore this job won't execute without one of them.

For 3. I follow the same code style as old one to make it more consistence.

For 4. `pipewire` is not only audio system but also video system. Therefore `pipewire` is used to be video system in jammy alongside with `pulseaudio` audio system. For jammy, `pw-cli` and `pactl` are both installed by default and will always use `pipewire` in your script.

Reply:
Above for 4 is for 5?

For 4. the echo message in the script is for manual test to notice user to make sure they are testing the right thing.

Reply:
For 2. I'm fine to remove it. However I prefer to comment it only to make user understand the limitation.

For echo message. If you know the `checkbox audio tests`, `pulseaudio` and `pipewire` well then it could be known from somewhere in the log.

Reply:
Please do remove it, the comment itself doesn't reveal useful information. 
Plainbox will skip the test if requires not meets, so user shouldn't try to debug in test-script layer.

Reply:
What if someone add a new job with this script and forget to add package requirement ?

Reply:
As per our discussion, this is a bit weird, keeping it for now but this is one possible way to follow this up